keps/sig-node/0014-runtime-class.md

```diff
@@ -0,0 +1,311 @@
+are too specific to a single runtime implementation should use `Parameters` rather than expanding
+the `RuntimeClassSpec`.
+
+**Becoming a general policy mechanism.** RuntimeClass should not be used a replacement for
+PodSecurityPolicy. The use cases for defining multiple RuntimeClasses for the same underlying
+runtime implementation should be extremely limited (generally only around updates & rollouts). To
+enforce this, no authorization or restrictions are placed directly on RuntimeClass use; in order to
+restrict a user to a specific RuntimeClass, you must use another policy mechanism such as
+PodSecurityPolicy.
+
+**Pushing complexity to the user.** RuntimeClass is a new resource in order to hide the complexity
+of runtime configuration from most users (aside from the cluster admin or provisioner). However, we
+are still side-stepping the issue of precisely defining specific types of runtimes like
+"Sandboxed". However, it is still up for debate whether precisely defining such runtime categories
+is even possible. RuntimeClass allows us to decouple this specification from the implementation, but
+it is still something I hope we can address in a future iteration through the concept of pre-defined
+or "conformant" RuntimeClasses.
+
+**Non-portability.** We are already in a world of non-portability for many features (see [examples
+of runtime variation](#examples-of-runtime-variation). RuntimeClasses actually address this issue by
+formally declaring supported features. Another issue is that pods need to refer to a RuntimeClass by
+name, which may not be defined in every cluster. This is something that can be addressed through
+pre-defined runtime classes (see previous risk), and/or by "fitting" pod requirements to compatible
+RuntimeClasses.
+
+## Graduation Criteria
+
+Alpha:
+
+- Everything described in the current proposal
+- Supported features and associated admission controller
+
+Beta:
+
+- Major runtimes support RuntimeClass
+- RuntimeClasses are configured in the E2E environment with test coverage of a non-legacy RuntimeClass
+- TBD
+
+## Implementation History
+
+- 2018-06-11: SIG-Node decision to move forward with proposal
+- 2018-06-19: Initial KEP published.
+
+## Appendix
+
+### Examples of runtime variation
+
+- Linux Security Module (LSM) choice - Kubernetes supports both AppArmor & SELinux options on pods,
+  but those are mutually exclusive, and support of either is not required by the runtime. The
+  default configuration is also not well defined.
+- Seccomp-bpf - Kubernetes has alpha support for specifying a seccomp profile, but the default is
+  defined by the runtime, and support is not guaranteed.
+- Windows containers - isolation features are very OS-specific, and most of the current features are
+  limited to linux. As we build out Windows container support, we'll need to add windows-specific
+  features as well.
+- Host namespaces (Network,PID,IPC) may not be supported by virtualization-based runtimes
```

Comment:
i was hoping that userns defaulting logic could be applied via runtimeclass as well for similar reasons.

/cc @vikaschoudhary16

Reply:
@derekwaynecarr userns feature will require a new runtimeclass because it is not backward compatible. If scheduling awareness is there, runtimeclasses will be more helpful in upgrades. New pods with userns runtimeclass will get steered to upgraded nodes. And it will be possible to have `hostUserNamespace=false` by default if `NodeLevelUserNamespace` is enabled. But for now, i think keeping `hostUserNamespace=true` as default makes more sense.

Reply:
I'd like to be careful about expanding the scope to setting backwards-incompatible defaults. I think options that are expressed in the PodSpec might be better off defaulted in cluster policy, even though our current cluster policy story (e.g. PodSecurityPolicy) is weak right now. I'm not saying RuntimeClass can't be used for that, I just don't think it's a usecase we should focus on right now. WDYT?

Reply:
+1 on both @vikaschoudhary16 and @tallclair said. More specifically, we might extend runtimeclass in the future for backward-incompatible features, but let's keep the alpha version's scope smaller. We can iterate it later.